### PR TITLE
rplidar_Modified_A2

### PR DESCRIPTION
**Pull request by luckywang95**
- Modified start_motor() & stop_motor() for start & stop motor for RPLidar A2 model.
- New function motor_setPWM for motor speed control.

### CODE REVIEW

**Comment by newpavlov** [comment]
Thank you! Though you could've committed changes directly into `rplidar.py`. This way it's easier to review them.

I've modified your code a bit, check changes if you are interested. It's mostly cleanups and compatibility changes for Python 2.

Modified code works with A1, but we need check it with A2 to be safe.

Also I've removed your changes to the `iter_measurments`, `iter_scans` and `clear_input`. Was there a particular reason for those changes?

**Comment by luckywang95** [comment]
I didn't make any modification to the functions you mentioned. At least in the version i downloaded.
I will update you after i tested it with my A2. I am currently still working on running my completed code on raspberry pi. 

btw, i have just completed a code similar to animate.py but i failed to do it by using FuncAnimation and iter_measurment because it keep giving me "Incorrect descriptor starting bytes" error. I think it is due to the start() in the iter_measurement(). Do you mind tell me how to perform FuncAnimation if i need to use the iter_measurment() instead of iter_scan()? However, I have completed a similar code but using different approach.

**Comment by newpavlov** [comment]
>I didn't make any modification to the functions you mentioned. At least in the version i downloaded.

Ah, so it was probably an older version of the module.

>because it keep giving me "Incorrect descriptor starting bytes" error.

Most probably it was due to the unread leftovers in the input buffer. Try to call `clear_input()` before starting `iter_scans()` or add correct shutdown in the end of your script.

`iter_measurments()` yields points instead of the whole scans, so I doubt you want to use it.

**Comment by luckywang95** [comment]
I think it is because i call iter_measurement() in my FuncAnimation()
`def lidar_plot(self):`
`    for measurement in lidar.iter_measurments_results_only():`
`lidar_plotting = animation.FuncAnimation(fig, lidar_plot, init_func = init,
                                             interval = 0, blit = True)`
so the scan has started but i still send command for get_health(), therefore, i always get the error. I dont know how to bypass that.
i use iter_measurement because i need the startBit. Instead of displaying the data constantly, i reduced the plot update speed, e.g. i update the plot only after 2 or 3 startBit is detected and i only store scan data before fig.canvas.update() so I have less data to transmit later.

**Comment by newpavlov** [comment]
You don't need start flag for that. You can simply write:
```Python
i = 0
for scan in lidar.iter_scans():
    if i == 10:
        process(scan)
        i = 0
    i += 1
```

Also before skipping scans, test speed of the new `animate.py`.

**Comment by luckywang95** [comment]
iter_scans() give u a list of distance, angle, and quality? how about when distance == 0? because now i abandoned points with distance == 0. by abandon these points, i can reduce the number of data transmitted to my MCU later.

for the new animate.py, the speed has improved but still not realt-time. By skipping only 2 scans and at max speed, i can get relatively smooth and real-time display. if at default speed, skipping 1 scans is enough.

I think we can send command such as get_health() but just either the program will read the response scan data or a mixture of health status and scan data.

**Comment by newpavlov** [comment]
I think RPLidar does not return measurements with zero distance or quality. But it should be checked.

Regarding `get_health()` I meant what module wasn't written with such use case in mind. I don't think it makes much sense to check health status constantly.

**Comment by luckywang95** [comment]
I did write the raw data to a txt file and analyze and it did give me 0 quality and 0 distance. at least for A2, it did. A1 didnt give data with 0 quality or 0 distance?

**Comment by newpavlov** [comment]
This is strange, because in the `iter_scan()` explicit filtering of points with zero quality implemented like this:
```Python
if data[1] > 0:
    scan.append(data[1:])
```
Maybe you used `iter_measurments()`? It does not perform any filtering and yields all points reported by the sensor, even invalid ones.

I think it's better to continue discussion not related to this pull request to the email. If you don't have any additions to it I will merge it.

**Comment by luckywang95** [comment]
Dont merge yet because i am able to start the motor but not able to stop the motor nor set_pwm. i am currently looking at the code.

File: rplidar.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def stop_motor(self):
         '''Stops sensor motor'''
         self.logger.info('Stoping motor')
         # For A1
-        self._serial_port.setDTR(True)
+        #self._serial_port.setDTR(True)  #here causing conflict with A2. DTR has to be always False to communicate.
         # For A2
         self.set_pwm(0)
```
Comment (newpavlov) [inline rplidar.py]: You can simply change order of `self._serial_port.setDTR(True)` and `self.set_pwm(0)`, no need to comment out it completely.

Comment (luckywang95) [inline rplidar.py]: yes. good idea also. but, although it is not necessary, user will not able to use set_pwm(duty_cycle) to start motor after they used stop_motor() to stop the motor?

Comment (newpavlov) [inline rplidar.py]: They will be able to call `start_motor()` which will call `setDTR(False)` and `set_pwm(DEFAULT_MOTOR_PWM)`. After that they will be able to change `pwm` if necessary.

File: rplidar.py [hunk 2]
```diff
@@ -155,10 +155,10 @@ def start_motor(self):
     def stop_motor(self):
         '''Stops sensor motor'''
         self.logger.info('Stoping motor')
-        # For A1
-        #self._serial_port.setDTR(True)  #here causing conflict with A2. DTR has to be always False to communicate.
         # For A2
         self.set_pwm(0)
+        # For A1
+        self._serial_port.setDTR(True)  #here causing conflict with A2. DTR has to be always False to communicate.
```
Comment (newpavlov) [inline rplidar.py]: Unnecessary comment left after last commit.

**Comment by newpavlov** [comment]
Thank you for your work!

I will need to make some changes (mainly documentation and support for express scan) before publishing new version to PyPI.

**Pull request merged**